### PR TITLE
Feat add recap view for pomodoro extention 

### LINKED ISSUE #11054: Stats for Pomodoro
### Extension

https://www.raycast.com/asubbotin/pomodoro

### Description

It'll be great to have some kind of record stats so I can know how many sessions daily, even weekly or monthly.

BTW, thank you for this plugin, it helps me a lot.

### Who will benefit from this feature?

All users.

### Anything else?

The report of https://pomofocus.io/ seems like a proper reference.

### PR DESCRIPTION
**Pull request by cchalop1**
## Description

I have had a recap page for pomodoro extention.

The goal is to have a page for resume all the data from the your previous pomodoro cycle.

I have had 3 informations:
- number of cycle completed
- number of time (in minutes) completed
- numbers of back to back (cycles with less than 5 min between tow)

I thinks we can add more informations in futures but we can make a other PR for that.

fix #11054 

## Screencast

Screenshot 2024-04-23 at 10 11 27

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by pernielsentikaer** [comment]
Hi @cchalop1 👋 

It looks good to me, @KELiON do you mind checking this?

**Comment by cchalop1** [comment]
Thanks for the typo !

File: extensions/pomodoro/src/stats-pomodoro-timer.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  let total = 0;
+  for (const interval of intervals) {
+    for (const part of interval.parts) {
+      if (part.endAt) {
+        total += interval.length;
+      }
+    }
+  }
+  return total;
+}
```
Comment (mikikiv) [inline extensions/pomodoro/src/stats-pomodoro-timer.tsx:28]: Should this be filtered to only count `interval.type === "focus"`?

Comment (pernielsentikaer) [reply extensions/pomodoro/src/stats-pomodoro-timer.tsx:28]: Can you help here @cchalop1

Comment (cchalop1) [reply extensions/pomodoro/src/stats-pomodoro-timer.tsx:28]: Yes you are right. I'm going to fix it.

**Comment by pernielsentikaer** [comment]
Just waiting for @KELiON do check it now 🙂

**Review by mikikiv — APPROVED** [review]

**Comment by pernielsentikaer** [comment]
I haven't tested it well yet, but we can maybe optimize `stats-pomodoro-timer.tsx` but using something like this which will:

* Avoid unnecessary iterations
* Use more descriptive variable names
* Use map, reduce, filter functions


```
import { usePromise } from "@raycast/utils";
import { getIntervalHistory } from "../lib/intervals";
import { Detail } from "@raycast/api";
import { Interval } from "../lib/types";

function calculateStats(intervals: Interval[]) {
  let completedCycles = 0;
  let totalFocusTime = 0;
  let backToBackCycles = 0;
  let prevEndAt = null;

  for (const interval of intervals) {
    for (const part of interval.parts) {
      if (part.endAt) {
        completedCycles++;
        totalFocusTime += interval.length;

        if (prevEndAt && part.startedAt - prevEndAt < 5 * 60 + 10) {
          backToBackCycles++;
        }

        prevEndAt = part.endAt;
      }
    }
  }

  return { completedCycles, totalFocusTime, backToBackCycles };
}

export default function StatsPomodoro() {
  const { data, isLoading } = usePromise(getIntervalHistory);
  const focusIntervals = data?.filter((interval) => interval.type === "focus") || [];

  const { completedCycles, totalFocusTime, backToBackCycles } = calculateStats(focusIntervals);

  const markdown = `# 🍅 Pomodoro Recap 🍅\n
  > 📊 Statistics of your pomodoro timer - all from the begin\n
   - You have completed **${completedCycles}** pomodoro cicle${completedCycles > 1 ? "s" : ""}. ✨\n
   - Total of **${totalFocusTime / 60}m** of focus time. ⏱️\n
   - Top Number of back to back pomodoro cicle${backToBackCycles > 1 ? "s" : ""}: **${backToBackCycles}**. 👑\n
    `;

  return <Detail isLoading={isLoading} markdown={markdown} />;
}
```

**Comment by cchalop1** [comment]
> I haven't tested it well yet, but we can maybe optimize `stats-pomodoro-timer.tsx` but using something like this which will:
> 
> * Avoid unnecessary iterations
> * Use more descriptive variable names
> * Use map, reduce, filter functions
> 
> ```
> import { usePromise } from "@raycast/utils";
> import { getIntervalHistory } from "../lib/intervals";
> import { Detail } from "@raycast/api";
> import { Interval } from "../lib/types";
> 
> function calculateStats(intervals: Interval[]) {
>   let completedCycles = 0;
>   let totalFocusTime = 0;
>   let backToBackCycles = 0;
>   let prevEndAt = null;
> 
>   for (const interval of intervals) {
>     for (const part of interval.parts) {
>       if (part.endAt) {
>         completedCycles++;
>         totalFocusTime += interval.length;
> 
>         if (prevEndAt && part.startedAt - prevEndAt < 5 * 60 + 10) {
>           backToBackCycles++;
>         }
> 
>         prevEndAt = part.endAt;
>       }
>     }
>   }
> 
>   return { completedCycles, totalFocusTime, backToBackCycles };
> }
> 
> export default function StatsPomodoro() {
>   const { data, isLoading } = usePromise(getIntervalHistory);
>   const focusIntervals = data?.filter((interval) => interval.type === "focus") || [];
> 
>   const { completedCycles, totalFocusTime, backToBackCycles } = calculateStats(focusIntervals);
> 
>   const markdown = `# 🍅 Pomodoro Recap 🍅\n
>   > 📊 Statistics of your pomodoro timer - all from the begin\n
>    - You have completed **${completedCycles}** pomodoro cicle${completedCycles > 1 ? "s" : ""}. ✨\n
>    - Total of **${totalFocusTime / 60}m** of focus time. ⏱️\n
>    - Top Number of back to back pomodoro cicle${backToBackCycles > 1 ? "s" : ""}: **${backToBackCycles}**. 👑\n
>     `;
> 
>   return <Detail isLoading={isLoading} markdown={markdown} />;
> }
> ```

Yes I agree with you, it is clearer if we put everything in the calculatesStats function. I push for change

**Comment by pernielsentikaer** [comment]
I hope I didn't break something since I didn't have any data to test with 🙂 Is this ready for another review?

**Comment by cchalop1** [comment]
I have tested with my local data and i work well. Yep it is ready !

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**